### ISSUE TITLE
Revisions to Notifier design

### ISSUE THREAD

**Issue by Chris-Hibbert** [Agoric/agoric-sdk#1089]
Had a conversation with @dtribble, he suggested the following:

 * the updatehandle should be a number rather than an object so it can be pass-by-copy. It doesn't matter if it's guessable, all that matters is that the current value is different from other recent values.
 * the `done` signal doesn't need to be a separate field. (For reducing messages sizes, reducing the number of fields is helpful). One possibility is that  done could be `undefined` until it's `true`, which at least means it wouldn't be serialized. Another is to check `updateHandle` for its final value instead, since that redundantly encodes the same state.
 * Another possibility would be to have updateHandle end up as `false` rather than `undefined`, to distinguish it from the initial (unknown) state. Then `getUpdateSince` could throws when updateHandle was false to encourage clients to check the final condition, rather than unintentionally allowing their code to continue requesting updates for dead channels.

**Comment by Chris-Hibbert** [comment]
@erights, please comment if this impacts #962

**Comment by erights** [comment]
> @erights, please comment if this impacts #962

Yes. Since I already need to make other non-independent changes to the notifier, I'm wondering if I should fold these into #962 ?

There are two semantic changes to notifier that I will make in #962. One in the start semantics and one in the end semantics.

# Start

Right now, if you ask for the value before the first update, you get `undefined`. You should get a promise which cannot be settled until an update happens. Likewise, an async iterator produces only an unresolved promise until it has a first value.

# End

Right now, the final "done" value is considered a normal part of the sequence of values, with the only thing special about it being that it is last. Both sync and async iterators treat the non-done values as being part of the normal sequence of values, that are observed in the same way. The final "done" value is normally non observed, and is in any case considered to be a different kind of thing that the values in the sequence. The best way to see this is in the coupling to JS syntax.

The generator (`function*) and async generator (`async function*) syntaxes produce generators, which are subtypes of iterators. Iterators are consumed by `for/of` loops and `for/await/of` loops. On generation, all `yield`ed values are non-done values in the sequence. The return value of the function becomes the done value of the iterator. The for loops bind the non-done values; but the done value of an iterator is *not even observable* by such for-loop code.

Since the whole point of notifiers is that more recent values are always better than older stale values. I expect many of our notifiers will be consumed by `for/await/of` loops, that will update, for example, a display to show the most recent contract state. The last meaningful contract state should therefore be the last non-done state. For a for loop consumer, the done-ness of an iterator is *only* a control signal for the for loop to exit.

Therefore, I plan to change the handling of `getUpdateSince` to not skip the last non-done state, even if the iterator is already done.

**Comment by erights** [comment]
My next speculative trick, probably after #962, is Nagling. A proxy that acts an a frozen pass-by-copy object does not need to commit to its values until observed, typically by marshal. Within the vat that's sending it out, the proxy's handler can maintain synchronous access to the notifier's current state. On observation, it can commit to that state, rather than what the state was when it was sent. It is a bit shocking, frankly, that this is possible. But it is impossible to prevent so we may as well make use of it.

For example, if a notifier state is updated several times within the same turn, the first update will cause the update record to be sent. But the update record, as observed by marshal, necessarily in a later turn, will observe an update record according to the last update in that turn.

Note that this special proxy behavior must be coded in SES, not Jessie. But the resulting notifier behavior can be used by Jessie code.

**Comment by Chris-Hibbert** [comment]
Thanks for the update, @erights. It does sound like you should fold this in to what you're working on. They're not at all independent.

**Comment by erights** [comment]
Worth mentioning again that generators, async generators, and `for/await/of` loops are all in SES but outside of Jessie. Thus, my stated expectation above that I expect for notifier uses to be consumed by `for/await/of` loops conflicts with my hope to get more people to program only in Jessie. It is a genuine tension.

**Comment by Chris-Hibbert** [comment]
That tension makes it sound like code inside Jessie would have to use `getUpdatesSince()` directly, while the generators would only be usable by code in SES. How much effort is it worth to get the generator version to work cleanly?

**Comment by erights** [comment]
> How much effort is it worth to get the generator version to work cleanly?

I still expect a lot, because of other libraries that already adapt or manipulate sync and async iterators. Node streams are directly adapted to async iterators. There are a number of iterator combinator libraries.

Also, we do support our uses using SES and, realistically, we should expect many of them to continue to use elements of SES outside Jessie. The strongest case of this is exactly the relevant one: `async/await`.

**Comment by dtribble** [comment]
WRT nagling, I note that batching within a turn is not sufficient. We want to batch across multiple messages queued to deliver to this vat, so that might even be beyond a crank. The technique of sending a message out to the kernel to be delivered "in order" which will naturally put it after other messages queued for this vat should give better nagling.

But the Proxy trick is cute.

**Comment by erights** [comment]
I don't understand nagling beyond a crank. My memory of nagling may be too simple. I haven't thought about it for decades. Could you step through a dirt simple example? Thanks.

> But the Proxy trick is cute.

;)

**Comment by Chris-Hibbert** [comment]
obsoleted by the current efforts: 

https://github.com/Agoric/agoric-sdk/pull/1340
https://github.com/Agoric/agoric-sdk/pull/1332
